File: solucoes/stack/expressoes/gabrielbergoc/expr.py
class Pilha:
    def __init__(self):
        self.__pilha = []

    def push(self, value):
        self.__pilha.append(value)

    def pop(self):
        return self.__pilha.pop()

    def is_empty(self):
        return len(self.__pilha) == 0


def abre(char):
    return char in ('(', '[', '{')


def validar(expressao):
    fecha = {'(': ')', '[': ']', '{': '}'}
    pilha = Pilha()

    for char in expressao:
        if abre(char):
            pilha.push(char)
            continue

        if pilha.is_empty():
            return False

        topo = pilha.pop()
        if fecha[topo] != char:
            return False

    return pilha.is_empty()


def main():
    T = int(input())

    for _ in range(T):
        expressao = input()
        valida = validar(expressao)
        print('S' if valida else 'N')


if __name__ == '__main__':
    main()

File: solucoes/stack/expressoes/gabrielbergoc/expr.cpp
#include <bits/stdc++.h>

using namespace std;

set<char> abre_ = {'(', '[', '{'};
unordered_map<char, char> fecha_ = {
    {'(', ')'},
    {'[', ']'},
    {'{', '}'},
};

bool abre(char c) {
    return abre_.find(c) != abre_.end();
}

char fecha(char c) {
    return fecha_[c];
}

bool validar(string expressao) {
    stack<char> fila;

    for (auto c : expressao) {
        if (abre(c)) {
            fila.push(c);
            continue;
        }

        if (fila.empty()) {
            return false;
        }

        char topo = fila.top();
        fila.pop();

        if (c != fecha(topo)) {
            return false;
        }

    }

    return fila.empty();
}

int main() {
    int T;
    cin >> T;

    for (int i = 0; i < T; i++) {
        string expressao;
        cin >> expressao;
        bool valida = validar(expressao);
        cout << (valida ? "S\n" : "N\n");
    }
}
